Implement initial polling for image processing

michizhou opened the pull request:
# Description

Implements image polling on the client side so that when using transformations which require processing and return a 423 status code, the image loads upon completion. This intends to use the `onError` function available for `next/future/image` components. For polling, the general approach is that when an error occurs, a simple fetch is made to the image URL. We will check the response and then continue polling until it's available if it's a 423 status code.

## Issue Ticket Number

Fixes #1 

## Type of change

<!-- Please select all options that are applicable. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Fix or improve the documentation
- [ ] This change requires a documentation update


# Checklist

<!-- These must all be followed and checked. -->

- [ ] I have followed the contributing guidelines of this project as mentioned in [CONTRIBUTING.md](/CONTRIBUTING.md)
- [ ] I have created an [issue](https://github.com/colbyfayock/next-cloudinary/issues) ticket for this PR
- [ ] I have checked to ensure there aren't other open [Pull Requests](https://github.com/colbyfayock/next-cloudinary/pulls) for the same update/change?
- [ ] I have performed a self-review of my own code
- [ ] I have run tests locally to ensure they all pass
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes needed to the documentation

colbyfayock commented:
hey @michizhou looking good so far

## Timeout

I think it would likely make sense to add a timeout to this so that it doesn't spam the image URL during polling which is likely to occur

Something i've done in the past to achieve this is along the lines of:
```
await (new Promise((resolve) => {
  setTimeout(() => resolve(), 500);
})
```

## Polling

It also didn't seem to work

![image](https://user-images.githubusercontent.com/1045274/197205876-6bc660f5-0873-4fb6-bf98-91d3b1c7773e.png)

Have you tested this locally? It doesnt seem to be polling for me

I suspect the reason may be because the 423 is triggering an error and you need the poll logic inside of the catch:
```
  try {
    let res = fetch(cloudinaryLoader({ ...imageOptions, options }, cldOptions));
    if (res.ok) {
      return res.json();
    } else if (res.status == 423) {
      return pollImage(imageOptions, options, cldOptions);
    }
  } catch (e) {
    // here
}
```

## Image Results

Finally, i'm wondering if the browser will "just update" when it sees it load successfully though im not sure that'll actually happen

Inside of the onError callback i wonder if we need to take the results and update some sort of internal state instance to force a reload of the component

this could potentially require a `key` property that would be different after loading

michizhou commented:
@colbyfayock Thank you for your comments. I'm not too sure how to test locally. Can you let me know how you tested locally so I can reproduce the development environment (and possible errors) on my end? I also added the timeout and modified the error image polling logic based on your comments. I would appreciate any further guidance as needed.

michizhou commented:
@colbyfayock Sorry to bother you, but I just wanted to follow up with you regarding my latest changes to this PR. Please let me know if additional changes are needed as well as how I can test out my changes locally.

colbyfayock commented:
@michizhou to test locally, set up your branch on your working environment

* run yarn dev at the top of the project
* open up two terminal tabs navigating one to next-cloudinary and one to docs and run yarn dev in both
* any time you make changes inside next-cloudinary, it will rebuild, and update teh docs site

colbyfayock commented:
unfortunately im not seeing any requests. i would recommend working to get this running locally to try

i need to create a solution for fully testing locally, but for now, you can upload an image to your cloudinary account, and open a Docs file like the background-removal.mdx file, and reference it in a component invocation 

on your cloudinary account, make sure to have the Cloudinary AI Background Removal add-on enabled. it has a free tier
